homeassistant/components/frontend/__init__.py

```diff
@@ -188,57 +261,71 @@ def add_manifest_json_key(key, val):
-    hass.http.register_view(BootstrapView)
+    import hass_frontend
+
     hass.http.register_view(ManifestJSONView)
 
-    if hass.http.development:
-        sw_path = "home-assistant-polymer/build/service_worker.js"
+    conf = config.get(DOMAIN, {})
+
+    frontend_path = hass_frontend.where()
+    repo_path = conf.get(CONF_FRONTEND_REPO)
+    is_dev = repo_path is not None
+
+    if is_dev:
+        hass.http.register_static_path("/home-assistant-polymer", repo_path)
+        sw_path = os.path.join(repo_path, "build/service_worker.js")
+        static_path = os.path.join(repo_path, 'hass_frontend')
+
     else:
-        sw_path = "service_worker.js"
+        sw_path = os.path.join(frontend_path, "service_worker.js")
+        static_path = frontend_path
 
-    hass.http.register_static_path("/service_worker.js",
-                                   os.path.join(STATIC_PATH, sw_path), False)
+    hass.http.register_static_path("/service_worker.js", sw_path, False)
     hass.http.register_static_path("/robots.txt",
-                                   os.path.join(STATIC_PATH, "robots.txt"))
-    hass.http.register_static_path("/static", STATIC_PATH)
+                                   os.path.join(frontend_path, "robots.txt"))
+    hass.http.register_static_path("/static", static_path)
 
     local = hass.config.path('www')
     if os.path.isdir(local):
         hass.http.register_static_path("/local", local)
 
-    index_view = hass.data[DATA_INDEX_VIEW] = IndexView()
+    index_view = IndexView(is_dev)
     hass.http.register_view(index_view)
 
-    # Components have registered panels before frontend got setup.
-    # Now register their urls.
-    if DATA_PANELS in hass.data:
-        for url_path in hass.data[DATA_PANELS]:
-            hass.http.app.router.add_route(
-                'get', '/{}'.format(url_path), index_view.get)
-            hass.http.app.router.add_route(
-                'get', '/{}/{{extra:.+}}'.format(url_path), index_view.get)
-    else:
-        hass.data[DATA_PANELS] = {}
+    @asyncio.coroutine
+    def finalize_panel(panel):
+        """Finalize setup of a panel."""
+        yield from panel.async_finalize(hass, repo_path)
+        panel.async_register_index_routes(hass.http.app.router, index_view)
 
-    if DATA_EXTRA_HTML_URL not in hass.data:
-        hass.data[DATA_EXTRA_HTML_URL] = set()
+    yield from asyncio.wait([
+        async_register_built_in_panel(hass, panel)
+        for panel in ('dev-event', 'dev-info', 'dev-service', 'dev-state',
+                      'dev-template', 'dev-mqtt', 'kiosk')])
 
-    for panel in ('dev-event', 'dev-info', 'dev-service', 'dev-state',
-                  'dev-template', 'dev-mqtt', 'kiosk'):
-        register_built_in_panel(hass, panel)
+    # Finalize registration of panels that registered before frontend was setup
+    # This includes the built-in panels from line above.
+    yield from asyncio.wait([finalize_panel(panel) for panel
```

Comment:
We should update `hass.data[DATA_FINALIZE_PANEL]` before this in case any panels register while this is waiting.